ecs-microservice/cognito.tf

```diff
@@ -85,5 +89,90 @@ resource "aws_ssm_parameter" "cognito-url" {
\ No newline at end of file
+# upload delegated cognito config to S3 bucket.
+# this will trigger the delegated cognito terraform pipeline and and apply the config.
+resource "aws_s3_bucket_object" "delegated-cognito-config" {
+  count = length(var.cognito_account_id)>0 ? 1 : 0
+  bucket = var.cognito_bucket
+  key    = "${length(var.cognito_env)>0 ? var.cognito_env : var.environment}/${local.current_account_id}/${var.name_prefix}-${var.service_name}.json"
+  acl    = "bucket-owner-full-control"
+
+  ## TODO maybe pull this out to a template to do more advanced conditional logic.
+  content = jsonencode({
+    # Configure resource server.
+    resource_server = {
+      name_prefix = "${var.name_prefix}-${var.service_name}"
+      identifier  = "${var.cognito_resource_server_identifier_base}/${var.service_name}"
+
+      scopes = [for key, value in var.resource_server_scopes : {
+        scope_name        = value.scope_name
+        scope_description = value.scope_description
+      }]
+    }
+
+    # Configure a user pool client
+    # TODO. this should be conditionally toggled.
+    user_pool_client = {
+      name_prefix     = "${var.name_prefix}-${var.service_name}"
+      generate_secret = true
+
+      allowed_oauth_flows                  = ["client_credentials"]
+      allowed_oauth_scopes                 = var.app_client_scopes
+      allowed_oauth_flows_user_pool_client = true
+    }
+  })
+  content_type = "application/json"
+}
+
+##
+# Read Credentials from Secrets Manager and set in microservice SSM config.
+# Store the md5 of the cognito config so that a change in md5/config
+# Will trigger a new update on dependent resources.
+#
+# Using workaround using time_sleep for async pipeline in cognito to complete
+# configuration of resource server and application client in delegated cognito.
+# The sleep wait will only occur when the dependent S3 file is updated
+# and during normal operation without changes it will not pause here.
+resource "time_sleep" "wait_for_credentials" {
+  count = length(var.cognito_account_id)>0 ? 1 : 0
+  create_duration = "300s"
+
+  triggers = {
+    config_hash = sha1(aws_s3_bucket_object.delegated-cognito-config[0].content)
+  }
+}
+
+# The client credentials that are stored in Central Cognito.
+data "aws_secretsmanager_secret_version" "microservice_client_credentials" {
+  depends_on = [aws_s3_bucket_object.delegated-cognito-config[0], time_sleep.wait_for_credentials[0]]
+  count = length(var.cognito_account_id)>0 ? 1 : 0
+  secret_id = "arn:aws:secretsmanager:eu-west-1:${var.cognito_account_id}:secret:${local.current_account_id}-${var.name_prefix}-${var.service_name}"
+}
+
+# Store client credentials from Central Cognito in SSM so that the microservice can read it.
+# TODO probably find a more suitable name/location for the parameter.
+resource "aws_ssm_parameter" "client_id" {
+  count = length(var.cognito_account_id)>0 ? 1 : 0
+  name      = "/${var.name_prefix}/config/${var.service_name}/client_id"
```

Comment:
Burde disse egentlig erstatte cognito-clientID og secret? https://github.com/nsbno/terraform-aws-trafficinfo/blob/cf2c2bd89ab378bc58c96982efcca9179b8b1441/ecs-microservice/cognito.tf#L51

Reply:
ja de skal det etterhvert. men kan ikke gjøre det nå, da knekker vi nok noen tjenester som bruker lokal cognito idag til credentials imellom seg.

Reply:
Ok, skjønner. Hvordan skal disse varaiablene da benyttes i mikrotjenestene? Trenger man dem her med en gang, eller kan de heller legges inn som cognito/clientId idet man gjør migreringen for hver tjeneste?

Reply:
Vi trenger dem ikke umiddelbart. Det viktigste er å få enabled opprettelse for alle tjenester i central cognito slik at vi ser at vi får samme oppsett der som vi har i vår lokale cognito. Deretter må vi fine-tune navnene på de her parametrene og oppsettet i tilgangskontrollen til micronaut. Fram til da vil ikke denne gjøre noe som helst forskjell for applikasjonene.

Reply:
I så fall tenker jeg det er best å vente med å legge til disse parameterene og fjerne dem herfra nå. Evt. gi dem et navn som tydelig sier at de hører til central cognito oppsett. Blir litt forvirrende å ha både client_id og clientId i param store